modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -0,0 +1,254 @@
+          case None => InvalidChain.raiseError[F, Agg]
+        }
+      }
+
+      getGenesisSnapshot.flatMap {
+        case (fullSnapshot, incrementalSnapshot) =>
+          go(incrementalSnapshot, GlobalSnapshotInfoV1.toGlobalSnapshotInfo(fullSnapshot.info))
+      }
+    }
+
+    def getGenesisSnapshot: F[(GlobalSnapshot, Signed[GlobalIncrementalSnapshot])] =
+      snapshotStorage
+        .readGenesis(lastFullGlobalSnapshotOrdinal)
+        .flatMap {
+          _.map(_.pure[F]).getOrElse {
+            fetchGenesis(lastFullGlobalSnapshotOrdinal)
+          }
+        }
+        .flatMap { genesis =>
+          val incrementalGenesisOrdinal = genesis.ordinal.next
+
+          snapshotStorage.readPersisted(incrementalGenesisOrdinal).flatMap {
+            case Some(snapshot) => (genesis.value, snapshot).pure[F]
+            case None           => fetchSnapshot(none[Hash], incrementalGenesisOrdinal).map((genesis, _))
+          }
+        }
+
+    private def fetchSnapshot(hash: Option[Hash], ordinal: SnapshotOrdinal): F[Signed[GlobalIncrementalSnapshot]] =
+      clusterStorage.getPeers
+        .map(_.toList)
+        .flatMap(Random[F].shuffleList)
+        .flatTap { _ =>
+          logger.info(s"Downloading snapshot hash=${hash.show}, ordinal=${ordinal.show}")
+        }
+        .flatMap { peers =>
+          type Success = Signed[GlobalIncrementalSnapshot]
+          type Agg = (List[Peer], Option[Signed[GlobalIncrementalSnapshot]])
+          type Result = Option[Success]
+
+          (peers, none[Signed[GlobalIncrementalSnapshot]]).tailRecM[F, Result] {
+            case (Nil, snapshot) => snapshot.asRight[Agg].pure[F]
+            case (peer :: tail, _) =>
+              p2pClient.globalSnapshot
+                .get(ordinal)
+                .run(peer)
+                .flatMap(_.toHashed[F])
+                .map(_.some)
+                .handleError(_ => none[Hashed[GlobalIncrementalSnapshot]])
+                .map {
+                  case Some(snapshot) if hash.forall(_ === snapshot.hash) => snapshot.signed.some.asRight[Agg]
+                  case _                                                  => (tail, none[Signed[GlobalIncrementalSnapshot]]).asLeft[Result]
+                }
+          }
+        }
+        .flatMap {
+          case Some(snapshot) => snapshot.pure[F]
+          case _              => CannotFetchSnapshot.raiseError[F, Signed[GlobalIncrementalSnapshot]]
+        }
+
+    def fetchGenesis(ordinal: SnapshotOrdinal): F[Signed[GlobalSnapshot]] =
+      clusterStorage.getPeers
+        .map(_.toList)
+        .flatMap(Random[F].shuffleList)
+        .flatTap { _ =>
+          logger.info(s"Downloading genesis snapshot ordinal=${ordinal}")
+        }
+        .flatMap { peers =>
+          type Success = Signed[GlobalSnapshot]
+          type Agg = (List[Peer], Option[Signed[GlobalSnapshot]])
+          type Result = Option[Success]
+
+          (peers, none[Signed[GlobalSnapshot]]).tailRecM[F, Result] {
```

Comment:
Since it's defined above?
```suggestion
          (peers, none[Success]).tailRecM[F, Result] {
```

Reply:
:+1: